src/Estimators/src/UkfState.cpp

```diff
@@ -0,0 +1,327 @@
+        dynamicsGroup->setParameter("sampling_time", state->m_pimpl->dT);
+
+        // create variable handler
+        std::string dynamicsName;
+        std::vector<double> covariances;
+        std::vector<std::string> dynamicsElements;
+        if (!dynamicsGroup->getParameter("name", dynamicsName))
+        {
+            log()->error("{} Unable to find the parameter 'name'.", logPrefix);
+            return nullptr;
+        }
+        if (!dynamicsGroup->getParameter("covariance", covariances))
+        {
+            log()->error("{} Unable to find the parameter 'covariance'.", logPrefix);
+            return nullptr;
+        }
+        if (!dynamicsGroup->getParameter("elements", dynamicsElements))
+        {
+            log()->error("{} Unable to find the parameter 'elements'.", logPrefix);
+            return nullptr;
+        }
+        state->m_pimpl->stateVariableHandler.addVariable(dynamicsName, covariances.size(), dynamicsElements);
+
+        std::string dynamicModel;
+        if (!dynamicsGroup->getParameter("dynamic_model", dynamicModel))
+        {
+            log()->error("{} Unable to find the parameter 'dynamic_model'.", logPrefix);
+            return nullptr;
+        }
+
+        std::shared_ptr<Dynamics> dynamicsInstance = RDE::DynamicsFactory::createInstance(dynamicModel);
+        if (dynamicsInstance == nullptr)
+        {
+            log()->error("{} The dynamic model '{}' has not been registered.", logPrefix, dynamicModel);
+            return nullptr;
+        }
+
+        dynamicsInstance->setSubModels(subModelList, kinDynWrapperList);
+
+        dynamicsInstance->initialize(dynamicsGroup);
+
+        // add dynamics to the list
+        state->m_pimpl->dynamicsList.insert({dynamicsName, dynamicsInstance});
+    }
+
+    // finalize estimator
+    if (!state->finalize(state->m_pimpl->stateVariableHandler))
+    {
+        log()->error("{} Unable to finalize the RobotDynamicsEstimator.", logPrefix);
+        return nullptr;
+    }
+
+    return state;
+}
+
+void RDE::UkfState::setUkfInputProvider(std::shared_ptr<const UkfInputProvider> ukfInputProvider)
+{
+    m_pimpl->ukfInputProvider = ukfInputProvider;
+}
+
+Eigen::MatrixXd RDE::UkfState::getNoiseCovarianceMatrix()
+{
+    return m_pimpl->covarianceQ;
+}
+
+System::VariablesHandler RDE::UkfState::getStateVariableHandler()
+{
+    return m_pimpl->stateVariableHandler;
+}
+
+void RDE::UkfState::propagate(const Eigen::Ref<const Eigen::MatrixXd>& cur_states, Eigen::Ref<Eigen::MatrixXd> prop_states)
+{
+    constexpr auto logPrefix = "[UkfState::propagate]";
+
+    // Check that everything is initialized and set
+    if (!m_pimpl->isFinalized)
+    {
+        log()->error("{} The ukf state is not well initialized.", logPrefix);
+        throw std::runtime_error("Error");
+    }
+
+    if (m_pimpl->ukfInputProvider == nullptr)
+    {
+        log()->error("{} The ukf input provider is not set.", logPrefix);
+        throw std::runtime_error("Error");
+    }
```

Comment:
This differs from the error-catching mechanism we used in the rest of the library. In particular, we mainly used an output boolean to indicate a failure, rather than an exception. I also understand that you needed to override this method, so there was not so much of an option in the first place.

With the risk of sounding like a broken record, I wonder if we could have something like
```cpp
bool advance(){
    try 
    {
        propagate(inputs);
    } 
    catch (std::runtime_error)
    {
        return false;
    }
    return true;
}
```

In this way, you would use this method as all the other methods in the library.

Reply:
I see your point, but this is not applicable here because the method `propagate` is used directly from the 
[bayes-filters-lib](https://github.com/robotology/bayes-filters-lib) library. So, I can't have an additional method (`advance`) as I don't have the possibility to call it.

Reply:
Ah, I see. So you never expect the user to call ``propagate``, right?

Reply:
> ever expect the user to call `propagate`

Yes, that’s right.

Reply:
Sounds good then!